pyomo/dae/contset.py

```diff
@@ -247,3 +248,54 @@ def construct(self, values=None):
         self._fe = sorted(self)
         timer.report()
 
+    def find_nearest_index(self, target, tolerance=None):
+        """ Returns the index of the nearest point in the 
+        :py:class:`ContinuousSet <pyomo.dae.ContinuousSet>`.
+
+        If a tolerance is specified, the index will only be returned
+        if the distance between the target and the closest point is
+        less than or equal to that tolerance. If the tolerance is less
+        than half the minimum spacing between points in the set, the
+        "closest-point-within-tolerance" will be unique.
+
+        Parameters
+        ----------
+        target : `float`
+        tolerance : `float` or `None`
+
+        Returns
+        -------
+        `float` or `None` 
+        """
+        lo = 1
+        hi = len(self) + 1
+        i = bisect.bisect_right(self, target, lo=lo, hi=hi)
+        # i is the index at which target should be inserted if it is to be
+        # right of any equal components. 
+
+        if i == lo:
+            # target is less than every entry of the set
+            nearest_index = i
+            delta = self[nearest_index] - target
+        elif i == hi:
+            # target is greater than or equal to every entry of the set
+            nearest_index = i-1
+            delta = target - self[nearest_index]
+        else:
+            # p_le <= target < p_g
+            # delta_left = target - p_le
+            # delta_right = p_g - target
+            # neighbors = {
+            #     delta_left: i-1,
+            #     delta_right: i,
+            #     }
+            # If delta_left == delta_right, delta_right will be kept as it is 
+            # added last. In this way, a tie goes to the index on the right.
+            neighbors = dict((abs(target - self[j]), j) for j in [i-1, i])
+            delta = min(neighbors)
+            nearest_index = neighbors[delta]
```

Comment:
Why make neighbors a dict?  If the key is the same, I cannot guarantee that `i` will (or will not) always replace `i-1`.  What about:
```python
delta, nearest_index = min((abs(target - self[j]), j) for j in [i-1, i])
```
You should note that this changes the handling for points exactly in the middle (ties now return the left point, not the right).

Reply:
Constructing a dict from a generator does not guarantee that the insertion order of the dict is the same as the generator? Regardless, your implementation is cleaner. I just didn't consider calling `min` on tuples.